### ISSUE TITLE
Should equalToJson treat null field and non-existent field the same?

### ISSUE THREAD

**Issue by yunj** [tomakehurst/wiremock#899]
## Bug reports
The request mapping is configured with the following request body.
```json
{
   "email":"email@hotmail.com",
   "homeAddress":{
      "line1":"1 Road",
      "suburb":"St",
      "postCode":"1071"
   },
   "mobile":{
      "subscriberNumber":"21123456",
      "countryCode":"61"
   }
}
```
**But when the actual request with null field, it fails the matching.**
```json
{
   "email":"email@hotmail.com",
   "homeAddress":{
      "line1":"1 Road",
      "line2":null,
      "suburb":"St",
      "postCode":"1071"
   },
   "mobile":{
      "subscriberNumber":"21123456",
      "countryCode":"61"
   }
}
```
equalToJson(xxx, true, true) is not perfect because we do want it failing if line2 does have value rather than null.

## Feature requests
From the restful PUT point of view, these two jsons are the same, so I think it should treat null field and non-existent field the same at this case.

**Comment by tomakehurst** [comment]
Please can you post the full stub mapping and requests that replicates this

**Comment by yunj** [comment]
Hi @tomakehurst,

Thanks very much for your reply. The stub mapping is as follows:
```json
{
   "id":"051f1c61-51ba-474f-8eef-715fa00ce805",
   "request":{
      "url":"/api/v1/my/contact",
      "method":"PUT",
      "headers":{
         "token":{
            "equalTo":"34c91881-986c-4381-ac81-5d2ca85a987a"
         }
      },
      "bodyPatterns":[
         {
            "equalToJson":"{\r\n  \"email\": \"email@hotmail.com\"\r\n  ,\"homeAddress\": {\r\n  \"line1\": \"1 Road\"\r\n}\r\n\r\n}\r\n"
         }
      ]
   },
   "response":{
      "status":204
   },
   "uuid":"051f1c61-51ba-474f-8eef-715fa00ce805"
}
````

**Comment by tomakehurst** [comment]
And the request(s)?

**Comment by yunj** [comment]
Hi @tomakehurst ,

Not sure if this is what you need? When doing a restful PUT request with the following body.
```json
{
   "email":"email@hotmail.com",
   "homeAddress":{
      "line1":"1 Road",
      "line2":null
   }
}
```

**Comment by yunj** [comment]
Hi @tomakehurst 

Does wiremock delegate the json comparison to the third-party library?

Thanks.

**Comment by tomakehurst** [comment]
Yes, it uses this: https://github.com/flipkart-incubator/zjsonpatch

**Comment by yunj** [comment]
Hi @tomakehurst ,

I'm wondering if you are going to have a look at this issue?

Thanks.

**Comment by yunj** [comment]
Hi @jaubuchon ,

it does not satisfy my requirement. As described in [here](https://github.com/tomakehurst/wiremock/issues/899#issue-303417198), we do want it failing if line2 does have value rather than null.

**Comment by nanditha167** [comment]
@yunj  was this issues resolved , Even I am facing the same issue .. @tomakehurst do you have any thoughts on it

**Comment by yunj** [comment]
> @yunj was this issues resolved , Even I am facing the same issue .. @tomakehurst do you have any thoughts on it

It remains unresolved.

**Comment by Blejwi** [comment]
Facing same issue, is there any workaround?

**Comment by tomakehurst** [comment]
I think this can't be changed without breaking backwards compatibility, and I also think that it's not always the case that null === field not present.

So, I think the solution is to add an extra flag to the `equalToJson` matcher telling it to treat the two as equivalent.

Marking this as a feature request. Happy to receive a PR if anyone feels motivated to create one.

**Comment by jonathansp** [comment]
any news?

**Comment by ohhyungsuh** [comment]
Hi @tomakehurst, happy 2025!  
I’d love to work on this issue — would that be ok?
Thank you!
